add child process termination functionality

northernSage opened the pull request:
- [x] extend `xprocess.py` to also handle child process termination
- [x] add tests
- [x] clean up
- [x] add change log 

Fixes  #12

northernSage commented:
A few things:

- The new process tree termination behaviour is exposed through the flag `kill_proc_tree` which can be optionally passed to `XProcessInfo.terminate` method, defaulting to `True` when not explicitly set. The reason behind this is allowing users that rely on the old behaviour of not properly ending all children procs (no clue why someone would use that, but I prefer to be on the safe side...) to be able to disable the new recursive termination if they so choose.

- test_server.test_shutdown_legacy was creating zombies when handling multi-process tests since the old `XProcessInfo.kill` is being used, so I opted for creating a separate `server-multiprocess.py` module and use it only on tests that use the current `XProcessInfo.terminate`

nicoddemus commented on CHANGELOG.rst:
```diff
@@ -0,0 +1,97 @@
+Unreleased
+
+- Added support for recursively terminating entire process tree with
+  :meth:`XProcessInfo._terminate_children()`
```
Not sure we should be documenting the internal function, we probably should make it clear that the default `XProcessInfo.terminate` has changed. How about:

```rst
- Now ``XProcessInfo.terminate`` will by default also terminate the entire process tree. This is safer as there's no risk of leaving lingering processes behind. If for some reason you need the previous behavior of only terminating the root process, ```kill_proc_tree=True`` to ``XProcessInfo.terminate``.
```

Also this is a new feature, so this should be `0.14` (technically it is a breaking change, but we are still in `0.X` so just bumping the minor number should be fine).

nicoddemus commented on xprocess.py:
```diff
@@ -25,7 +25,24 @@ def __init__(self, path, name):
+        for p in alive:
+            p.kill()
+
+    def terminate(self, kill_proc_tree=True):
```
Let's make this a keyword only argument:

```python
def terminate(self, **kwargs):
    kill_proc_tree = kwargs.pop("kill_proc_tree", True)
    if kwargs:
        raise TypeError("unknown keyword arguments: {}".format(kwargs.keys()))      
    ...
```

This of course can be made nicer once we drop Python 2.7:
```python
def terminate(self, *, kill_proc_tree=True):
    ...
```

northernSage replied:
Done. Since you mentioned python 2.7, I was considering opening an issue to discuss the possibility of dropping it. I would gladly start working on a PR for that if someone give me a green light (or is there a reason to keep it for the time being that I am not seeing?).

nicoddemus replied:
> Done. Since you mentioned python 2.7, I was considering opening an issue to discuss the possibility of dropping it. I would gladly start working on a PR for that if someone gives me the green light (or is there a reason to keep it for the time being that I am not seeing?).

We usually wait until supporting 2.7 becomes a burden or gets into the way of implementing new functionality. Until that time comes and the maintenance burden is reasonable, I think it is best to keep it working. But to be clear: if there's some new functionality that can only be implemented in Python 3+, I'm glad to use that as an excuse to drop 2.7. 👍 

> Finally, I was tinkering around with tox and had to recreate my venv for testing. I forgot to reinstall pre-commit hooks and only noticed after pushing changes (silly me), so the commit history is a bit polluted. I can rebase this if necessary.

Thanks, but don't worry, we've been squash+merging your latest PRs as their changes are self contained in my opinion. 👍

northernSage replied:
It makes sense, thanks for explaining.

nicoddemus commented on example/server-multiprocess.py:
```diff
@@ -0,0 +1,39 @@
+import sys
+
+
+response = os.environ.get("RESPONSE", "1").encode("utf8")
```
Unused?

northernSage replied:
Oops.

nicoddemus reviewed (changes requested):
Thanks @northernSage! This looks great, left a few comments though. 👍

nicoddemus reviewed (approved):
Great work!

nicoddemus commented:
@northernSage I would like to thank you to your recent contributions to the project!

I've sent you an invitation to become a member. If you accept, would you also like to prepare the next release? The procedure is the same as xdist's: https://github.com/pytest-dev/pytest-xdist/blob/master/RELEASING.rst (we should also copy that text over to this repository).

nicoddemus commented:
Hmm it seems `test_clean_shutdown` is flaky:

```
=================================== FAILURES ===================================
_____________________________ test_clean_shutdown ______________________________

xprocess = <xprocess.XProcess object at 0x7f795494afd0>

    def test_clean_shutdown(xprocess):
        proc_names = ["server", "server2", "server3"]
        all_children = [
            psutil.Process(xprocess.getinfo(name).pid).children() for name in 
        ]
        children_pids = []
        for proc_children in all_children:
            assert len(proc_children) >= 1
            children_pids += [c.pid for c in proc_children]
        for name in proc_names:
            xprocess.getinfo(name).terminate()
        for pid in children_pids:
>           assert not psutil.pid_exists(pid)
E           assert not True
E            +  where True = <function pid_exists at 0x7f7954a91290>(2741)
E            +    where <function pid_exists at 0x7f7954a91290> = psutil.pid_exists
```

We should probably use a "wait" strategy in there.

northernSage commented:
Thanks, I'm glad I could help a little and hope I can keep contributing even more. I will get started on the release asap (probably will be out in the next few days). Regarding the flaky test, I'll take a more careful look at it later, but this seems odd since `_terminate_children` already `psutil.wait_procs(children, timeout=timeout)`. I would guess that the termination might be timing out and the assertion is probably failing because `for p in alive: p.kill()` isn't done yet.

nicoddemus commented:
> Thanks, I'm glad I could help a little and hope I can keep contributing even more.

Sure thing, glad to help you onboard and thanks for the interest in the project!

> Regarding the flaky test, I'll take a more careful look at it later, but this seems odd since _terminate_children already psutil.wait_procs(children, timeout=timeout). I would guess that the termination might be timing out and the assertion is probably failing because for p in alive: p.kill() isn't done yet and processes are still lingering for a short delay after SIGKILL is issued.

Perhaps we need a new round of waiting after `kill()` then?

```diff
        children = proc.children(recursive=True)
        for p in children:
            p.terminate()
        _, alive = psutil.wait_procs(children, timeout=timeout)
        for p in alive:
            p.kill()
+       _, alive = psutil.wait_procs(children, timeout=timeout)
+       if alive:
+           raise psutil.Error
```

Raising `psutil.Error` here will make `terminate()` return `-1` (failed to terminate).

(Another nice improvement in Python 3 would change that return from an `int` to an `IntEnum` hehehe)

northernSage commented:
Okay, after some research (mostly psutil docs) I have changed `XProcessInfo.terminate` considerably (also added your suggestions) and now tests seem to be consistent and taking much less time to run. I won't merge this right now since I ended up altering more than I had planned and a second opinion on this would be nice.

nicoddemus commented on xprocess.py:
```diff
@@ -47,17 +47,20 @@ def terminate(self, **kwargs):
-        _, alive = psutil.wait_procs(kill_list, timeout=timeout)
-        if alive:
+        try:
+            parent = psutil.Process(self.pid)
```
Looks good!

To be ultra-safe, how about moving this line outside this try/finally, into a separate try/except:

```python
try:
    parent = psutil.Process(self.pid)
except NoSuchProcess:
    return 0
```

This is a bit more "beg for forgiveness" than "ask for permission" approach taken in line 47. That line can then just be changed to:

```python
if not self.pid:
    return 0
```

WDYT?

northernSage replied:
Nice catch. It is (as you said) unlikely to happen, but definitely possible if the process is externally terminated, for example, due to some exceptional circumstance like extreme resource starvation where it would get forcefully terminated by the kernel.

nicoddemus commented on xprocess.py:
```diff
@@ -47,17 +47,20 @@ def terminate(self, **kwargs):
         if not self.pid or not self.isrunning():
```
(not related to this line, but GH only let's me comment on the diff)

I only now noticed "setting ``kill_proc_tree`` flag", but I believe a more natural wording would be "setting ``kill_proc_tree`` keyword-only parameter".

northernSage replied:
Done

nicoddemus commented on xprocess.py:
```diff
@@ -47,17 +47,20 @@ def terminate(self, **kwargs):
         if not self.pid or not self.isrunning():
             return 0
         timeout = 20
```
While we are updating this method, how about also turning this into a keyword-only parameter instead?

northernSage replied:
Yes, being able to adjust how long to wait on process termination based on specificities of context (e.g what kind of external process your are running) makes sense for the user.

northernSage commented:
Well, unless someone has more suggestions I believe we should be ready for the next release :tada:

nicoddemus commented:
Great @northernSage, thanks again! Feel free to go ahead and prepare the release! 🎉
